src/dgisim/encoding/encoding_plan.py

```diff
@@ -261,6 +262,18 @@ def perspective_version(self, pid: Pid) -> Self:
             self._flip_cache = new_self
         return self._flip_cache
 
+class LazyEncodingPlan(EncodingPlan):
+    """
+    A lazy version of EncodingPlan that does not encode anything, returning None.
+
+    It can accelerate env.step() if you don't need the encoding instantly.
+    """
+    def encode(self, game_state: "GameState", perspective:Pid):
```

Comment:
Please update type-hint (return type) for both `.encode()` methods and methods that called `.encode()` as appropriate. (replace relevant `list[int]` with `None | list[int]`) (I believe the only ones affected are in  `LinearEnv`)

Reply:
It's reasonable to update type-hint for `LinearEnv`. However, updating the `.encode()` methods will violate the static type checking, for example, code in `__name__=="__main__"` section. It assumes that the return value is a list, and indeed the return value is a list, with no possibility of None.

A similar issue was discussed [here](https://github.com/microsoft/pylance-release/issues/4950). In my opinion it's not appropriate because for `EncodingPlan`, its `encode()` method actually returns `list[int]` instead of `None` and for `LazyEncodingPlan`, its `encode()` method indeed returns `None`.

Reply:
Updating `.encode()` does break static type checking for certain parts, and that is why we need to update those parts as well to fix the issue. If you read more about the project, you may find many `assert isinstance(item, SomeClass)` to pass static type checking.

The point is, as `EncodingPlan.encode()` promises a return type of `list[int]`, all subclasses of `EncodingPlan` needs to follow it as well, otherwise, users may be calling `encoding_plan.encode(...)`, and get `None` instead because `LazyEncodingPlan` is passed in instead of `EncodingPlan`. If we ensure both `.encode() -> None | list[int]`, then users are aware that they may actually get `None` before running the code and get an error.

Reply:
A simple example is:

```py
def some_function() -> EncodingPlan:
    return LazyEncodingPlan()

encoding_plan: EncodingPlan = some_function()
encoded_state = encoding_plan.encode(state, pid)
print(len(encoded_state))
```

This piece of code passes static type checking in code editors, but will break at runtime because encoded_state is `None` instead of `list[int]`.

Reply:
This made me rethink the code structure. I think the simplest way is to have LazyEncodingPlan return an empty list `[]`. This way, compatibility is maintained with the least amount of code. 

Going off on a tangent, I think the core issue is that I want to reuse other methods from `EncodingPlan`, modifying only the `encode` method. In reality, `EncodingPlan` and `LazyEncodingPlan` are not actually in a parent-child relationship. I want to respect semantics, but I violated the Liskov Substitution Principle.